### PR TITLE
chore: update CUDA.Dockerfile to use multistage builds

### PR DESCRIPTION
**Pull request by matifali**
1. Updates the Dockerfile to use a multistage build
2. Use Intel MKL instead of OpenBLAS
3. Reduced the Final image size from 17+ GB to ~12 GB

### CODE REVIEW

File: CUDA.Dockerfile [hunk 1]
```diff
@@ -1,10 +1,8 @@
-# Build argumnets
-ARG CUDA_VER=11.8.0
-ARG UBUNTU_VER=22.04
```
Comment (matifali) [inline CUDA.Dockerfile:3]: I have removed these ARGs. We can only build with 11.8.0, and aihwkit does not build with CUDA 12.x versions.

File: CUDA.Dockerfile [hunk 2]
```diff
@@ -40,26 +93,20 @@ RUN groupadd -g ${GROUPID} ${USERNAME} && \
...
+# Install Python packages
+RUN pip install --no-cache-dir --no-warn-script-location \
+    matplotlib
 
-# Copy the source code inside to image and change to the source directory
-COPY . ./aihwkit
-WORKDIR /home/${USERNAME}/aihwkit
+# Environment variables
+ENV PATH /home/${USERNAME}/.local/bin:$PATH
+ENV LD_LIBRARY_PATH /opt/intel/oneapi/mkl/latest/lib/intel64:${LD_LIBRARY_PATH}
```
Comment (matifali) [inline CUDA.Dockerfile:109]: This makes sure that MKL libs are in the path.

File: docs/source/advanced_install.rst [hunk 3]
```diff
@@ -257,9 +257,6 @@ The image can be built via::
     --build-arg USERID=(id -u $USER) \
     --build-arg GROUPID=(id -g $USER) \
     --build-arg CUDA_ARCH=${CUDA_ARCH} \
-    --build-arg CUDA_VER=11.8.0 \
-    --build-arg UBUNTU_VER=22.04 \
-    --build-arg PYTORCH_PIP_URL=https://download.pytorch.org/whl/cu117 \
```
Comment (matifali) [inline docs/source/advanced_install.rst:262]: Removed as the image is hard-coded to cuda 11.8.

File: CUDA.Dockerfile (hunk 1 above)
Comment (matifali) [reply CUDA.Dockerfile:3]: Edit: It builds fine, but while running requires some `cudablas11.so` files that are nit available with CUDA 12.1.

Comment (matifali) [reply CUDA.Dockerfile:3]: I will update

Comment (matifali) [reply CUDA.Dockerfile:3]: pushed a fix

**Comment by matifali** [comment]
Hi @maljoras it is ready for review.

**Review by maljoras — APPROVED** [review]
Many thanks @matifali !

**Pull request merged**